pkg/ext-proc/scheduling/scheduler.go

```diff
@@ -5,24 +5,74 @@ import (
 	"fmt"
 	"math/rand"
 
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 	klog "k8s.io/klog/v2"
 
 	"ext-proc/backend"
 )
 
+const (
+	// TODO Consider making this configurable.
+	kvCacheThreshold = 80
+)
+
 var (
+	allowAllFilter = &filter{
+		name:   "noop",
+		filter: toFilterFunc(allowAllPredicate),
+	}
+
 	defaultFilter = &filter{
+		name:          "critical request",
+		filter:        toFilterFunc(criticalRequestPredicate),
+		nextOnSuccess: criticalRequestFilter,
+		nextOnFailure: sheddableRequestFilter,
+	}
+
+	// The goal for scheduling critical requests is to minimize the latency. The heuristic is to
+	// pick a server with least "load" (KV Cache), which typically yields lower latency.
+	// Heuristics for scheduling critical requests:
+	criticalRequestFilter = &filter{
 		name:   "least queuing",
 		filter: leastQueuingFilterFunc,
 		nextOnSuccessOrFailure: &filter{
-			name:   "lora affinity",
-			filter: toFilterFunc(loraAffinityPredicate),
+			name:   "low cost LoRA",
+			filter: toFilterFunc(lowLoRACostPredicate),
 			nextOnSuccessOrFailure: &filter{
 				name:   "least KV cache percent",
 				filter: leastKVCacheFilterFunc,
 			},
 		},
 	}
+
+	// The goal for scheduling sheddable requests is to optimize for throughput while reducing
+	// queuing, and leave low load (KV cache) servers to serve critical requests.
+	sheddableRequestFilter = &filter{
+		// When there is at least one model server that's not queuing requests, and still has KV
+		// cache below a certain threshold, we consider this model server has capacity to handle
+		// a sheddable request without impacting critical requests.
+		name:   "has capacity for sheddable requests",
+		filter: toFilterFunc(noQueueAndLessThanKVCacheThresholdPredicate(kvCacheThreshold)),
+		nextOnSuccess: &filter{
+			name:   "most KV cache percent",
+			filter: mostKVCacheFilterFunc,
```

Comment:
I will also compare different combinations such as using leastKVCache

Reply:
Sending the request to the highest kv-cache server may actually have a counter effect on throughput because it may increase prefill interruptions and kv-cache preemptions. We need to have a limit, and this limit is the highest kv-cache that doesn't cause evictions (80%).

Reply:
The limit is in the noQueueAndLessThanKVCacheThresholdPredicate. The intuition was by preferring higher kv-cache we leave room for critical requests. However my benchmarking didn't really show a big difference in the critical request latencies, so I just changed to the same filter as the critical request (prefer low kv cache).